Transactions (#174)
dbos-inc/dbos-docs: docs/tutorials/transaction-tutorial.md

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> Use _transaction functions_ to read and write from your database. A transaction function may contian multiple queries, punctuated by local logic, and executes as a single [database transaction](https://en.wikipedia.org/wiki/Database_transaction). 

Reviewer: ```suggestion
Use _transaction functions_ to read and write from your database. A transaction function may contain multiple queries as well as TypeScript business logic and executes as a single [database transaction](https://en.wikipedia.org/wiki/Database_transaction). 
```

Transaction functions must be annotated with the [`@Transaction`](../api-reference/decorators#transaction) decorator and must have a [`TransactionContext`](../api-reference/contexts#transactioncontextt) as their first argument.
As with other DBOS functions, inputs and outputs must be serializable to JSON.

The [`TransactionContext`](../api-reference/contexts#transactioncontextt) provides a `.client` field you can use to transactionally interact with the database, so you don't need to worry about managing database connections.
The [`TransactionContext`](../api-reference/contexts#transactioncontextt) provides a `.client` field you can use to interact with the database, so you don't need to worry about managing connections.
DBOS supports [Knex.js](./using-knex.md), [Drizzle](./using-drizzle.md), [TypeORM](./using-typeorm.md), and [Prisma](./using-prisma.md) clients as well as raw SQL.
You can configure which client to use in your [`dbos-config.yaml`](../api-reference/configuration.md) file.

> Knex is used by default.

Reviewer: To be honest, I think we recommend using [Drizzle `sql` templates](https://orm.drizzle.team/docs/sql) for raw SQL...

> See below for an example of a write and a read transaction with each option. Note we recommend marking reads as `{readOnly: true}` for faster performance.

Reviewer: ```suggestion
Here are examples of a write and a read transaction function in each client. Note we recommend marking reads as `{readOnly: true}` for faster performance.
```

> See below for an example of a write and a read transaction with each option. Note we recommend marking reads as `{readOnly: true}` for faster performance.

Reviewer: Make the `readOnly` thing a tip box?

Author: Done

```javascript
interface GreetingRecord {
  name: string;
  note: string;
}

export class Greetings {
  //...
  @Transaction()
  static async insertGreeting(ctxt: TransactionContext<Knex>, gr: GreetingRecord) {

>     await ctxt.client('greetings').insert(gr);

Reviewer: If it's not too much trouble, might be nice to annotate each RO function with its return type so it's clearer what they're actually doing.

Author: done

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

Use _transaction functions_ to read and write from your database. A transaction function may contain multiple queries as well as TypeScript business logic and executes as a single [database transaction](https://en.wikipedia.org/wiki/Database_transaction). 

Transaction functions must be annotated with the [`@Transaction`](../api-reference/decorators#transaction) decorator and must have a [`TransactionContext`](../api-reference/contexts#transactioncontextt) as their first argument.
As with other DBOS functions, inputs and outputs must be serializable to JSON.

The [`TransactionContext`](../api-reference/contexts#transactioncontextt) provides a `.client` field you can use to interact with the database, so you don't need to worry about managing connections.
DBOS supports [Knex.js](./using-knex.md), [Drizzle](./using-drizzle.md), [TypeORM](./using-typeorm.md), and [Prisma](./using-prisma.md) clients as well as raw SQL.
You can configure which client to use in your [`dbos-config.yaml`](../api-reference/configuration.md) file.
Knex is the default and we recommend using `Knex.raw()` for raw SQL.

> :::tip

Reviewer: Maybe move the tip box below the examples?

<TabItem value="drizzle" label="Drizzle">

```javascript
export const GreetingRecord = pgTable('greetings', {
  name: text('name'),
  note: text('note'),
});

export class Greetings {
  //..
  @Transaction()
  static async insertGreeting(ctxt: TransactionContext<NodePgDatabase>, name: string, note: string) {
    await ctxt.client.insert(GreetingRecord).values({name: name, note: note});
  }

  @Transaction({ readOnly:true })

>   static async getGreetings(ctxt: TransactionContext<NodePgDatabase>): Promise<{name: string | null, note: string | null}[]> {

Reviewer 2: Nit: maybe define a type like other ORMs do:

interface GreetingRecord {
  name: string;
  note: string;
}

Author: In other ORMs it was a requirement - I had to create the type or else the linter would complain. With Drizzle, it's not really a requirement. Some people may see it as a benefit? So that's why I didn't make one here... WDYT?

Reviewer: Don't make one, Drizzle gets its types from your schema (and return types are just whatever you asked for, of course).

<TabItem value="prisma" label="Prisma">

```javascript
//Model specified in prisma/schema.prisma:
//
//model GreetingRecord {
//  @@map("greetings") 
//  greeting_id Int @id @default(autoincrement()) //Note: Prisma requires at least one primary key
//  name String
//  note String
//}

import { PrismaClient } from "@prisma/client"; // Use the generated Prisma client

export class Greetings {
  //...
  @Transaction()
  static async insertGreeting(ctxt: TransactionContext<PrismaClient>, name: string, note: string) {
    await ctxt.client.greetingRecord.create({
      data: {
        name: name,
        note: note
      },
    });
  }

  @Transaction({ readOnly:true })

>   static async getGreetings(ctxt: TransactionContext<PrismaClient>): Promise<{greeting_id: number, name:string, note:string}[]> {

Reviewer 2: You can directly use the defined GreetingRecord from PrismaClient. Example: https://github.com/dbos-inc/dbos-demo-apps/blob/main/bank/bank-backend/src/workflows/txnhistory.workflows.ts#L16

```ts
import { AccountInfo, PrismaClient, TransactionHistory } from "@prisma/client";
```
